File: SpreadsheetPython/Tester.py
import datetime
import Spreadsheet
import Cell


def tester_valid(cond, func_name, class_name):
    if cond:
        print(f"passed {func_name} for {class_name}")
    else:
        print(f"failed {func_name} for {class_name}")


""" tests for cell """


def test_cell_get_set_value():
    c = Cell.Cell()
    c.set_value("hel")
    tester_valid(c.get_value() == "hel", "get_value", "Cell")


def test_cell_get_set_color():
    c = Cell.Cell()
    c.set_color("white")
    tester_valid(c.get_color() == Cell.colors["white"], "get_color", "Cell")


def test_cell_value():
    c = Cell.Cell()
    c.set_value("-1")
    try:
        c.to_int()
    except:
        print("failed to_int for Cell, exception")
    else:
        if c.to_int() == -1:
            print("passed to_int for Cell")
        else:
            print("failed to_int for Cell")


    c.set_value("5.4")
    try:
        c.to_float()
    except:
        print("failed to_float for Cell, exception")
    else:
        if c.to_float() == 5.4:
            print("passed to_float for Cell")
        else:
            print("failed to_float for Cell")

    c.set_value("09/02/2023")
    try:
        c.to_date()
    except:
        print("failed to_date for Cell, exception")
    else:
        if str(c.to_date()) == "2023-02-09":
            print("passed to_date for Cell")
        else:
            print("failed to_date for Cell")


def test_cell_reset():
    c = Cell.Cell()
    c.set_value("hello")
    c.reset()
    tester_valid(c.get_value() == "", "reset", "Cell")


""" tests for spreadsheet """


def test_spreadsheet_set_get():
    c = Cell.Cell("hello")
    sp = Spreadsheet.Spreadsheet(2, 2)
    sp.set_cell_at(0, 0, c)
    sp.set_cell_at(1, 1, "hi")
    tester_valid(sp.get_cell_at(0, 0).get_value() == "hello", "get_cell_at(cell)", "Spreadsheet")
    tester_valid(sp.get_cell_at(1, 1).get_value() == "hi", "get_cell_at(str)", "Spreadsheet")


def test_spreadsheet_add_row():
    sp = Spreadsheet.Spreadsheet(2, 2)
    sp.set_cell_at(1, 1, "hi")
    sp.add_row(1)
    tester_valid(sp.get_cell_at(2, 1).get_value() == "hi", "add_row", "Spreadsheet")


def test_spreadsheet_remove_row():
    sp = Spreadsheet.Spreadsheet(3, 2)
    sp.set_cell_at(1, 0, "hi")
    sp.remove_row(1)
    tester_valid(sp.get_cell_at(1, 0).get_value() == "hi", "remove_row", "Spreadsheet")


def test_spreadsheet_add_column():
    sp = Spreadsheet.Spreadsheet(2, 2)
    sp.set_cell_at(1, 1, "hi")
    sp.add_column(1)
    tester_valid(sp.get_cell_at(1, 2).get_value() == "hi", "add_column", "Spreadsheet")


def test_spreadsheet_remove_column():
    sp = Spreadsheet.Spreadsheet(2, 3)
    sp.set_cell_at(1, 2, "hi")
    sp.remove_column(1)
    tester_valid(sp.get_cell_at(1, 1).get_value() == "hi", "remove_column", "Spreadsheet")


def test_spreadsheet_swap_rows():
    sp = Spreadsheet.Spreadsheet(2, 3)
    sp.set_cell_at(1, 2, "hi")
    sp.swap_rows(0, 1)
    tester_valid(sp.get_cell_at(0, 2).get_value() == "hi", "swap_rows", "Spreadsheet")


def test_spreadsheet_swap_columns():
    sp = Spreadsheet.Spreadsheet(2, 3)
    sp.set_cell_at(1, 2, "hi")
    sp.swap_columns(0, 2)
    tester_valid(sp.get_cell_at(1, 0).get_value() == "hi", "swap_columns", "Spreadsheet")


test_cell_get_set_value()
test_cell_get_set_color()
test_cell_value()
test_cell_reset()

test_spreadsheet_set_get()
test_spreadsheet_add_row()
test_spreadsheet_remove_row()
test_spreadsheet_add_column()
test_spreadsheet_remove_column()
test_spreadsheet_swap_rows()
test_spreadsheet_swap_columns()

File: SpreadsheetCPP/Version1/Tester.cpp
#include <iostream>
#include "Spreadsheet.cpp"


void testCell() {
	Cell c1 = Cell("hello", Color::red);

	Cell c = Cell();
	c.setValue("23");
	c.setColor(Color::red);

	if(c.getColor() == Color::red) {
		std::cout << "ok set/get color" << std::endl;
	} else {
		std::cout << "! color" << std::endl;
	}
	if(c.getValue() == "23") {
		std::cout << "ok for string" << std::endl;
	} else {
		std::cout << "! string" << std::endl;
	}
	c.reset();
	if(c.getValue() == "") {
		std::cout << "ok. reseted" << std::endl;
	} else {
		std::cout << "not reseted" << std::endl;
	}
}

void testCellValueInt() {
	Cell c = Cell("23", Color::red);
	if(c.toInt() == 23) {
		std::cout << "ok for int 23" << std::endl;
	} else {
		std::cout << "! int 23" << std::endl;	
	}
	c.setValue("023");
	if(c.toInt() == 23) {
		std::cout << "ok for int 023" << std::endl;
	} else {
		std::cout << "! int 023" << std::endl;	
	}
	c.setValue("-23");
	if(c.toInt() == -23) {
		std::cout << "ok for int -23" << std::endl;
	} else {
		std::cout << "! int -23" << std::endl;	
	}
	c.setValue("+23");
	if(c.toInt() == 23) {
		std::cout << "ok for int +23" << std::endl;
	} else {
		std::cout << "! int +23" << std::endl;	
	}
}

void testCellValueDouble() {
	Cell c = Cell("23.4");
	if(c.toDouble() == 23.4) {
		std::cout << "ok for double 23.4" << std::endl;
	} else {
		std::cout << "! double 23.4" << std::endl;
	}
	c.setValue("023.4");
	if(c.toDouble() == 23.4) {
		std::cout << "ok for double 023.4" << std::endl;
	} else {
		std::cout << "! double 023.4" << std::endl;	
	}
	c.setValue("-23.4");
	if(c.toDouble() == -23.4) {
		std::cout << "ok for double -23.4" << std::endl;
	} else {
		std::cout << "! double -23.4" << std::endl;	
	}
	c.setValue("+23.4");
	if(c.toDouble() == 23.4) {
		std::cout << "ok for double +23.4" << std::endl;
	} else {
		std::cout << "! double +23" << std::endl;	
	}
}

void testCellValueDate() {
	Cell c = Cell("08/02/2023");
	if(c.toDate() == Date(8, 02, 2023)) {
		std::cout << "ok for date 1" << std::endl;
	} else {
		std::cout << "! date 1" << std::endl;
	}
}

void testSpreadsheetSetGet() {
	Spreadsheet sp = Spreadsheet(3, 4);
	Cell cl = Cell("23");
	sp.setCellAt(0, 0, "hello");
	sp.setCellAt(0, 1, cl);
	if(sp.getCellAt(0, 1) == cl) {
		std::cout << "ok set/get" << std::endl;
	} else {
		std::cout << "! get/set" << std::endl;
	}
	try {
		sp.setCellAt(10, 10, "world");
	} catch(...) {
		std::cout << "ok out of range setCellAt 1" << std::endl;
	}
	try {
		sp.setCellAt(10, 10, cl);
	} catch(...) {
		std::cout << "ok out of range setCellAt 2" << std::endl;
	}
}

void testSpreadsheetAddRemove() {
	Spreadsheet sp = Spreadsheet(3, 4);  //size is 3x4
	sp.addRow(1);                        //size is 4x4, new row is sp[2][i]
	try {
		sp.setCellAt(3, 3, "hello");
	} catch(...) {
		std::cout << "! failed addRow" << std::endl;
	}
	sp.addRow(3);                        //size is 5x4, new row is sp[4][i]
	try { 
		sp.addRow(10); 
	} catch(...) {
		std::cout << "ok for out of range addRow" << std::endl;
	}
	sp.removeRow(0);                     //size is 4x4, removed row is sp[0][i]
	try { 
		sp.removeRow(10); 
	} catch(...) {
		std::cout << "ok for out of range removeRow" << std::endl;
	}
	sp.addColumn(1);                     //size is 4x3, removed column is sp[i][1]
	try { 
		sp.removeColumn(10); 
	} catch(...) {
		std::cout << "ok for out of range removeColumn" << std::endl;
	}
}

void testSpreadsheetSwap() {
	Spreadsheet sp = Spreadsheet(3, 4);   //3 and 4 are not indexes
	Cell cl = Cell("hello");
	sp.setCellAt(0, 0, cl);
	sp.swapRows(0, 2);                    //0 and 2 are indexes
	if(sp.getCellAt(2, 0) == cl) {
		std::cout << "ok swapRows" << std::endl;
	} else {
		std::cout << "! swapRows" << std::endl;
	}
	sp.swapColumns(0, 1);
	if(sp.getCellAt(2, 1) == cl) {
		std::cout << "ok swapColumns" << std::endl;
	} else {
		std::cout << "! swapColumns" << std::endl;
	}
}


int main() {
	testCell();
	testCellValueInt();
	testCellValueDouble();
	testCellValueDate();

	testSpreadsheetSetGet();
	testSpreadsheetAddRemove();
	testSpreadsheetSwap();

	return 0;
}


